### ISSUE TITLE
Fatal error while decoding

### ISSUE THREAD

**Issue by richardhenry** [fumoboy007/msgpack-swift#2]
Hi, thanks for open sourcing this library!

I was testing it out and ran into a reproducible out of bounds error in MessageReader in `readByte()` on the line `return message[currentIndex]`.  Here is a concise version of the situation that causes this crash:

```
struct Event: Decodable {
    let t: EventType
    let ts: String
    let p: Data
}

let data = try MessagePackDecoder().decode([Event].self, from: data)
```

In case it is relevant, EventType looks like this:

```
enum EventType: Int, Codable {
    case example = 1
    // ...
}
```

This crash only seems to occur with a specific input, and it's not sensitive, so I'm happy to share it here as base64:

```
k4OhdAGidHPZXjMwOjE3MDI5NTU5NDA4NDYtMDoxNToxNzAyNjgxMDM2NjkzLTA6MzE6MTcwMjYxMzM5MDE4MC0wOjEwOjE3MDI2MTI1ODEwODYtMDo0MDoxNzAyNjEyOTIyNDk3LTChcMUBGYqheQGhZ6tJRTg5OTNlVHlES6F02SAxNzNhYzM2YzRjNmY0YmRlYTdhNWUwOThlNjE2MDg4YqFt2SBkNzFjMjI1N2E2MzA0ZTk4Yjc2MmMzYzkyN2I5NjRiZqF1qzczdWxtaWxqY3lDoWTNBd2hcsKhY8RWMwohBXQbaUup8eei5dBi4PL+oyzOZE2rSPjbtidTB78gZ0M3EAwYBCIkB12k1F9eQeDo0KiluQ3oEkce/pQI4pKoeMJAOqNuLfY7VtzS6x46XhTUz12heM8AAAGMgBV77KFzxEDSVdIBrLXch8jDHcqIcC1damYRxXi6/BdkFKMmzS66ztFLaP6qlKQPzgR+sBjK3Y6Db+0uSP9Mh0KupQNkM66Kg6F0AaJ0c9leMzA6MTcwMjk1NTk0MTU0MC0wOjE1OjE3MDI2ODEwMzY2OTMtMDozMToxNzAyNjEzMzkwMTgwLTA6MTA6MTcwMjYxMjU4MTA4Ni0wOjQwOjE3MDI2MTI5MjI0OTctMKFwxQEZiqF5AaFnq0lFODk5M2VUeURLoXTZIDE3M2FjMzZjNGM2ZjRiZGVhN2E1ZTA5OGU2MTYwODhioW3ZIDU0NmU2MDM0MWE5YTQ5NTc4YWVkZWYzYmZhNDI1Y2ZjoXWrNzN1bG1pbGpjeUOhZM0F3aFywqFjxFYzCiEFdBtpS6nx56Ll0GLg8v6jLM5kTatI+Nu2J1MHvyBnQzcQDRgEIiQxppzkexZF6sGS+K44MRsY78674EM3ByeytsTp717x+FwKxRW4il0htAfrPqF4zwAAAYyAFX6foXPEQJ0hVj6vNIzc/Sa4bg3hSSyvJAJtv4xHHO9vX3pPXVveGvZVFZSVy1VTIIC5YbKBIxxpJrF4EtyPoF/4LKPQn46DoXQBonRz2V4zMDoxNzAyOTU1OTQyMTMwLTA6MTU6MTcwMjY4MTAzNjY5My0wOjMxOjE3MDI2MTMzOTAxODAtMDoxMDoxNzAyNjEyNTgxMDg2LTA6NDA6MTcwMjYxMjkyMjQ5Ny0woXDFARmKoXkBoWerSUU4OTkzZVR5REuhdNkgMTczYWMzNmM0YzZmNGJkZWE3YTVlMDk4ZTYxNjA4OGKhbdkgZDFmNWE4ZWQwNGRkNDY1YjgwZTRmM2U3MmJlMGJjZGKhdas3M3VsbWlsamN5Q6FkzQXdoXLCoWPEVjMKIQV0G2lLqfHnouXQYuDy/qMszmRNq0j427YnUwe/IGdDNxAOGAQiJFUrQMtR2ZpSpiRShGOBkMB7IALBJIU06w2iKJ6ZJbcM3jRgr9Ie5T0Fp6pfoXjPAAABjIAVgO6hc8RALpT3g361rf4sj0ArSbAUCGFJ8WVseMFbJxzXVmJUlWNnuwDIJdUBxD+4IObdLfunAdn76niBsIwrYHrPqwLJiA==
```

I am able to decode this data using MessagePacker and the msgpack-c library.

Let me know if you're able to reproduce this issue. Thanks!

**Comment by fumoboy007** [comment]
@richardhenry Thanks for reporting this issue!

I am not able to reproduce it. I ran the following test case 100 times in both debug and release modes on an Intel Mac:
```swift
func testCrash() throws {
   enum EventType: Int, Codable {
       case example = 1
   }
   struct Event: Decodable {
       let t: EventType
       let ts: String
       let p: Data
   }

   let data = Data(base64Encoded: "k4OhdAGidHPZXjMwOjE3MDI5NTU5NDA4NDYtMDoxNToxNzAyNjgxMDM2NjkzLTA6MzE6MTcwMjYxMzM5MDE4MC0wOjEwOjE3MDI2MTI1ODEwODYtMDo0MDoxNzAyNjEyOTIyNDk3LTChcMUBGYqheQGhZ6tJRTg5OTNlVHlES6F02SAxNzNhYzM2YzRjNmY0YmRlYTdhNWUwOThlNjE2MDg4YqFt2SBkNzFjMjI1N2E2MzA0ZTk4Yjc2MmMzYzkyN2I5NjRiZqF1qzczdWxtaWxqY3lDoWTNBd2hcsKhY8RWMwohBXQbaUup8eei5dBi4PL+oyzOZE2rSPjbtidTB78gZ0M3EAwYBCIkB12k1F9eQeDo0KiluQ3oEkce/pQI4pKoeMJAOqNuLfY7VtzS6x46XhTUz12heM8AAAGMgBV77KFzxEDSVdIBrLXch8jDHcqIcC1damYRxXi6/BdkFKMmzS66ztFLaP6qlKQPzgR+sBjK3Y6Db+0uSP9Mh0KupQNkM66Kg6F0AaJ0c9leMzA6MTcwMjk1NTk0MTU0MC0wOjE1OjE3MDI2ODEwMzY2OTMtMDozMToxNzAyNjEzMzkwMTgwLTA6MTA6MTcwMjYxMjU4MTA4Ni0wOjQwOjE3MDI2MTI5MjI0OTctMKFwxQEZiqF5AaFnq0lFODk5M2VUeURLoXTZIDE3M2FjMzZjNGM2ZjRiZGVhN2E1ZTA5OGU2MTYwODhioW3ZIDU0NmU2MDM0MWE5YTQ5NTc4YWVkZWYzYmZhNDI1Y2ZjoXWrNzN1bG1pbGpjeUOhZM0F3aFywqFjxFYzCiEFdBtpS6nx56Ll0GLg8v6jLM5kTatI+Nu2J1MHvyBnQzcQDRgEIiQxppzkexZF6sGS+K44MRsY78674EM3ByeytsTp717x+FwKxRW4il0htAfrPqF4zwAAAYyAFX6foXPEQJ0hVj6vNIzc/Sa4bg3hSSyvJAJtv4xHHO9vX3pPXVveGvZVFZSVy1VTIIC5YbKBIxxpJrF4EtyPoF/4LKPQn46DoXQBonRz2V4zMDoxNzAyOTU1OTQyMTMwLTA6MTU6MTcwMjY4MTAzNjY5My0wOjMxOjE3MDI2MTMzOTAxODAtMDoxMDoxNzAyNjEyNTgxMDg2LTA6NDA6MTcwMjYxMjkyMjQ5Ny0woXDFARmKoXkBoWerSUU4OTkzZVR5REuhdNkgMTczYWMzNmM0YzZmNGJkZWE3YTVlMDk4ZTYxNjA4OGKhbdkgZDFmNWE4ZWQwNGRkNDY1YjgwZTRmM2U3MmJlMGJjZGKhdas3M3VsbWlsamN5Q6FkzQXdoXLCoWPEVjMKIQV0G2lLqfHnouXQYuDy/qMszmRNq0j427YnUwe/IGdDNxAOGAQiJFUrQMtR2ZpSpiRShGOBkMB7IALBJIU06w2iKJ6ZJbcM3jRgr9Ie5T0Fp6pfoXjPAAABjIAVgO6hc8RALpT3g361rf4sj0ArSbAUCGFJ8WVseMFbJxzXVmJUlWNnuwDIJdUBxD+4IObdLfunAdn76niBsIwrYHrPqwLJiA==".data(using: .utf8)!)!

   let event = try MessagePackDecoder().decode([Event].self, from: data)
   print(event)
}
```

Which Swift compiler version, OS, and hardware are you using to reproduce this?

**Comment by richardhenry** [comment]
Well, I hate to say it, but I can't reproduce it now either. I'll spend some more time trying to figure out what happened here and reopen the issue if I'm able to figure it out!

**Comment by richardhenry** [comment]
I ended up just using a different MessagePack library for the last few months to avoid this issue, but I wanted to come back to this and I have a consistent repro now. It seems like it's related to calling `MessagePackDecoder.decode` again using Data that was allocated by a previous call to `MessagePackDecoder.decode`.

Here's a test case that reproduces this bug:

``` swift
import XCTest
import MessagePack

struct Outer: Codable {
    var data: Data
}

struct Inner: Codable {
    var number: Int
}

final class MessagePackCrashTests: XCTestCase {
    func testDoesNotCrash() throws {
        // This test does not crash.
        
        let inner = Inner(number: 5)
        let event = Outer(data: try MessagePackEncoder().encode(inner))
        _ = try MessagePackDecoder().decode(Inner.self, from: event.data)
    }

    func testCrashesEveryTime() throws {
        // This test crashes every time for me.

        let inner = Inner(number: 5)
        let event = Outer(data: try MessagePackEncoder().encode(inner))
        let encodedData = try MessagePackEncoder().encode(event)

        let decodedEvent = try MessagePackDecoder().decode(Outer.self, from: encodedData)
        _ = try MessagePackDecoder().decode(Inner.self, from: decodedEvent.data)  // <- Crashes here
    }
}
```

This is what I'm seeing:

<img width="652" alt="Screenshot 2024-05-28 at 12 25 22 AM" src="https://github.com/fumoboy007/msgpack-swift/assets/63480/062e3394-9103-40c1-8b1b-f7b8aa7d2813">

<img width="707" alt="Screenshot 2024-05-28 at 12 26 00 AM" src="https://github.com/fumoboy007/msgpack-swift/assets/63480/bdd09564-dd34-4bbb-842e-f4f2772e6182">

The test is running with "My Mac" as a target (which is a MacBookPro18,2 with Apple Silicon).

**Comment by richardhenry** [comment]
In the test case that doesn't crash, the data looks like this in the debugger:

<img width="349" alt="Screenshot 2024-05-28 at 12 38 09 AM" src="https://github.com/fumoboy007/msgpack-swift/assets/63480/17811167-8746-4f79-b7a3-a7c3178de598">

In the test case that does crash, it looks like this:

<img width="345" alt="Screenshot 2024-05-28 at 12 37 57 AM" src="https://github.com/fumoboy007/msgpack-swift/assets/63480/07a2ace8-3555-41cc-8411-1ec1035eef0e">

If I copy the data before I use it for the second decode call by calling `decodedEvent.data.withUnsafeBytes { Data($0) }`, the crash goes away.

**Comment by richardhenry** [comment]
This commit fixes the issue:

https://github.com/richardhenry/msgpack-swift/commit/51862567d727ed733c2026cd8c93a16997a0d73d

Maybe there's a better way to implement this change?

I'm still not totally understanding why this is happening - it seems like the data that is backing the slice has been deallocated, but I'm not sure why.

**Comment by fumoboy007** [comment]
@richardhenry Thanks for helping narrow down the issue!

The crash is due to a [precondition](https://github.com/apple/swift-foundation/blob/92aef643c99c3d5bcc272d5ad77362060e41a178/Sources/FoundationEssentials/Data/Data.swift#L982) failure when subscripting `MessageReader.message`. Strange that the failure message does not show up.

The bug is that `MessageReader` assumes that `Data.startIndex` is always zero. However, this may not be true for a slice.

Similar to [this](https://forums.swift.org/t/data-subscript/57195) Swift forum post, I was misled by the return type of the subscript operator. Unlike, say, `Array` whose subscript operator returns `ArraySlice` to make it clear that a slice is being returned, the subscript operator of `Data` returns `Data`, so I wrongly assumed that the collection would have a zero `startIndex`.

**Comment by richardhenry** [comment]
Aha, makes sense! Thanks for fixing.

**Comment by fumoboy007** [comment]
The fix has been released in [version 2.0.2](https://github.com/fumoboy007/msgpack-swift/releases/tag/2.0.2).
